src/librustc_mir/transform/uniform_array_move_out.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This pass converts move out from array by Subslice and
+// ConstIndex{.., from_end: true} to ConstIndex move out(s) from begin
+// of array. It allows detect error by mir borrowck and elaborate
+// drops for array without additional work.
+//
+// FIXME: convert to Subslice back for performance reason
+// FIXME: integrate this transformation to the mir build
+
+use rustc::ty;
+use rustc::ty::TyCtxt;
+use rustc::mir::*;
+use transform::{MirPass, MirSource};
+use util::patch::MirPatch;
+
+pub struct UniformArrayMoveOut;
+
+impl MirPass for UniformArrayMoveOut {
+    fn run_pass<'a, 'tcx>(&self,
+                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          _src: MirSource,
+                          mir: &mut Mir<'tcx>) {
+        let mut patch = MirPatch::new(mir);
+        for (bb, block) in mir.basic_blocks().iter_enumerated() {
+            for (i, stmt) in block.statements.iter().enumerate() {
+                let location = Location { block: bb, statement_index: i };
+                if let StatementKind::Assign(ref dst_place, ref rval) = stmt.kind {
+                    if let Rvalue::Use(ref operand) = *rval {
+                        if let Operand::Move(ref src_place) = operand {
+                            self.process_move_assign(
+                                tcx, mir, location, dst_place, src_place, &mut patch);
+                        }
+                    }
+                }
+            }
+        }
+        patch.apply(mir);
+    }
+}
+
+impl UniformArrayMoveOut {
+    fn process_move_assign<'a, 'tcx>(&self,
+                                     tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                     mir: &Mir<'tcx>,
+                                     location: Location,
+                                     dst_place: &Place<'tcx>,
+                                     src_place: &Place<'tcx>,
+                                     patch: &mut MirPatch<'tcx>) {
+        if let Place::Projection(ref proj) = *src_place {
+            let place_ty = proj.base.ty(mir, tcx).to_ty(tcx);
+            if let ty::TyArray(_t, size_o) = place_ty.sty {
+                let opt_size = size_o.val.to_const_int().and_then(|v| v.to_u64());
+                if let Some(size) = opt_size {
+                    let size = size as u32;
+                    match proj.elem {
+                        ProjectionElem::Subslice{from, to} => {
```

Comment:
Example comment:

---

Before we have:

```
X = move Y[3..5]
```

we want to transform to the following patch:

```
- X = move Y[3..5]
+ tmp0 = move Y[3];
+ tmp1 = move Y[4];
+ X = [tmp0, tmp1];
```